# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [1]:
# Import the appropriate libraries
import pandas as pd
import numpy as np
import matplotlib.mlab as mlab
import matplotlib
from nltk.metrics import edit_distance
import missingno as msno
import pandas_profiling as pp

# Visualization Imports
import matplotlib.pyplot as plt
import seaborn as sns
color = sns.color_palette()
get_ipython().run_line_magic('matplotlib', 'inline')
import plotly.offline as py
py.init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.tools as tls
import plotly.express as px
import numpy as np


from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

#Print multiple outputs from 1 cell


In [2]:
# Create two new dataframes
#Set the first column as index

north_df=pd.read_csv('NorthCallCenter.csv', index_col=[0])
south_df=pd.read_csv('SouthCallCenter.csv', index_col=[0])


In [3]:
pd.set_option('display.max_columns', None)

In [4]:
north_df.head()


Branch     Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
0  North        Complaint    9:00 AM             Incoming     B  Brent   NO   
1  North    Sales Support   11:00 AM             Incoming     A    Cam   NO   
2  North    Sales Support   11:00 AM             Incoming     B   Todd   NO   
3  North    Sales Support    4:00 PM             Incoming     B  Brent   NO   
4  North  Product Support   12:00 PM             Incoming     B  Brent   NO   

   Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  Incoming Wait Time  \
0          0          1             9                  2                 1.0   
1          0          0            11                  1                 1.0   
2          0          0             9                  3                 1.0   
3          0          0            11                  2                 1.0   
4          0          0             8                  2                 1.0   

   During Call Wait Time  Calls  
0                      4      1  
1                      3      1  
2                      3      1  
3                      4      1  
4                      4      1

In [5]:
south_df.head()

Branch   Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
0  South  Sales Support   10:00 AM             Incoming     D   Kate   NO   
1  South  Sales Support   10:00 AM             Incoming     C   Eric   NO   
2  South  Sales Support   10:00 AM             Incoming     C  Susan   NO   
3  South  Sales Support   10:00 AM             Incoming     C  Alice   NO   
4  South  Sales Support   12:00 PM             Incoming     C  Sandy   NO   

   Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  Incoming Wait Time  \
0          0          0             9                  1                 1.0   
1          0          0             8                  2                 1.0   
2          0          0            10                  2                 1.0   
3          0          0            12                  1                 1.0   
4          0          0             8                  3                 1.0   

   During Call Wait Time  Calls  
0                      2      1  
1                      4      1  
2                      4      1  
3                      3      1  
4                      3      1

In [6]:
north_df.shape

(245, 14)

In [7]:
south_df.shape

(314, 14)

In [8]:
north_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 245 entries, 0 to 244
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Branch                 245 non-null    object 
 1   Call Purpose           245 non-null    object 
 2   Time Block             245 non-null    object 
 3   Incoming or Outgoing   245 non-null    object 
 4   Queue                  245 non-null    object 
 5   Rep ID                 245 non-null    object 
 6   Sale                   245 non-null    object 
 7   Lost Call              245 non-null    int64  
 8   Abandoned              245 non-null    int64  
 9   Busy Minutes           245 non-null    int64  
 10  Not Ready Minutes      245 non-null    int64  
 11  Incoming Wait Time     82 non-null     float64
 12  During Call Wait Time  245 non-null    int64  
 13  Calls                  245 non-null    int64  
dtypes: float64(1), int64(6), object(7)
memory usage: 28.7+ KB


In [9]:
south_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 314 entries, 0 to 313
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Branch                 314 non-null    object 
 1   Call Purpose           314 non-null    object 
 2   Time Block             314 non-null    object 
 3   Incoming or Outgoing   314 non-null    object 
 4   Queue                  314 non-null    object 
 5   Rep ID                 314 non-null    object 
 6   Sale                   314 non-null    object 
 7   Lost Call              314 non-null    int64  
 8   Abandoned              314 non-null    int64  
 9   Busy Minutes           314 non-null    int64  
 10  Not Ready Minutes      314 non-null    int64  
 11  Incoming Wait Time     126 non-null    float64
 12  During Call Wait Time  314 non-null    int64  
 13  Calls                  314 non-null    int64  
dtypes: float64(1), int64(6), object(7)
memory usage: 36.8+ KB


In [10]:
north_df.columns

Index(['Branch', 'Call Purpose', 'Time Block', 'Incoming or Outgoing', 'Queue',
       'Rep ID', 'Sale', 'Lost Call', 'Abandoned', 'Busy Minutes',
       'Not Ready Minutes', 'Incoming Wait Time', 'During Call Wait Time',
       'Calls'],
      dtype='object')

In [11]:
south_df.columns

Index(['Branch', 'Call Purpose', 'Time Block', 'Incoming or Outgoing', 'Queue',
       'Rep ID', 'Sale', 'Lost Call', 'Abandoned', 'Busy Minutes',
       'Not Ready Minutes', 'Incoming Wait Time', 'During Call Wait Time',
       'Calls'],
      dtype='object')

# Exploratory Data Analysis

**Time to do some EDA! In the process of learning more about the two datasets, make sure you answer the following questions:**

1. How many reps are in each branch?
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?
3. What is the number of calls taken for each time block?

In [12]:
# Put your EDA code here

In [13]:
#north_df.nlargest(5, 'Incoming Wait Time')

**How many reps are in each branch?**

In [14]:
north_df['Rep ID'].nunique()

9

In [15]:
north_df['Rep ID'].describe()

count       245
unique        9
top       Brent
freq         37
Name: Rep ID, dtype: object

In [16]:
north_df['Rep ID'].value_counts()

Brent     37
Lilly     30
Todd      29
Duke      29
Joe       26
Xander    24
Cam       24
Amanda    23
Andy      23
Name: Rep ID, dtype: int64

In [17]:
south_df['Rep ID'].nunique()

11

In [18]:
south_df['Rep ID'].describe()

count      314
unique      11
top       Eric
freq        35
Name: Rep ID, dtype: object

In [19]:
south_df['Rep ID'].value_counts()

Eric      35
Randy     33
Sandy     32
Helga     31
Kate      29
George    29
Karl      28
Josh      26
Susan     26
Sharon    25
Alice     20
Name: Rep ID, dtype: int64

**What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?**

In [20]:
north_df.describe()

Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  245.000000  245.000000    245.000000         245.000000   
mean     0.020408    0.028571      9.987755           1.914286   
std      0.141681    0.166940      1.403980           0.771416   
min      0.000000    0.000000      8.000000           1.000000   
25%      0.000000    0.000000      9.000000           1.000000   
50%      0.000000    0.000000     10.000000           2.000000   
75%      0.000000    0.000000     11.000000           3.000000   
max      1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count           82.000000             245.000000  245.0  
mean             3.048780               2.967347    1.0  
std              1.456382               0.843827    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

In [21]:
south_df.describe()

Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  314.000000  314.000000     314.00000         314.000000   
mean     0.035032    0.012739      10.05414           1.910828   
std      0.184154    0.112324       1.40751           0.860016   
min      0.000000    0.000000       8.00000           1.000000   
25%      0.000000    0.000000       9.00000           1.000000   
50%      0.000000    0.000000      10.00000           2.000000   
75%      0.000000    0.000000      11.00000           3.000000   
max      1.000000    1.000000      12.00000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          126.000000             314.000000  314.0  
mean             3.000000               3.076433    1.0  
std              1.308434               0.819424    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

In [22]:
north_df[['Busy Minutes','Not Ready Minutes','Incoming Wait Time', 'During Call Wait Time']].mean()

Busy Minutes             9.987755
Not Ready Minutes        1.914286
Incoming Wait Time       3.048780
During Call Wait Time    2.967347
dtype: float64

In [23]:
south_df[['Busy Minutes','Not Ready Minutes','Incoming Wait Time', 'During Call Wait Time']].mean()

Busy Minutes             10.054140
Not Ready Minutes         1.910828
Incoming Wait Time        3.000000
During Call Wait Time     3.076433
dtype: float64

**What is the number of calls taken for each time block?**

In [24]:
north_df['Time Block'].value_counts()

12:00 PM    45
10:00 AM    43
2:00 PM     35
3:00 PM     30
4:00 PM     28
11:00 AM    25
5:00 PM     17
1:00 PM     17
9:00 AM      5
Name: Time Block, dtype: int64

In [25]:
south_df['Time Block'].value_counts()

12:00 PM    75
10:00 AM    56
3:00 PM     43
11:00 AM    31
2:00 PM     30
5:00 PM     26
4:00 PM     25
1:00 PM     23
9:00 AM      5
Name: Time Block, dtype: int64

In [26]:
north_df[["Time Block", "Calls"]].groupby("Time Block").sum()

Calls
Time Block       
10:00 AM       43
11:00 AM       25
12:00 PM       45
1:00 PM        17
2:00 PM        35
3:00 PM        30
4:00 PM        28
5:00 PM        17
9:00 AM         5

In [27]:
south_df[["Time Block", "Calls"]].groupby("Time Block").sum()

Calls
Time Block       
10:00 AM       56
11:00 AM       31
12:00 PM       75
1:00 PM        23
2:00 PM        30
3:00 PM        43
4:00 PM        25
5:00 PM        26
9:00 AM         5

### EDA Results

Make note of your results here!

**1. How many reps are in each branch?**



**North Branch:9**

Brent,
Lilly,
Duke, 
Todd,
Joe, 
Xander,
Cam,
Amanda,
Andy,

**South Branch:11**

Eric,
Randy,
Sandy,
Helga,
George,
Kate,
Karl,
Susan,
Josh,
Sharon,
Alice, 


**2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?**



**North Branch Average:**

Busy Minutes-             9.987755


Not Ready Minutes-        1.914286


Incoming Wait Time-       3.048780


During Call Wait Time-    2.967347


**South Branch Average:**

Busy Minutes-             10.054140


Not Ready Minutes-         1.910828


Incoming Wait Time-        3.000000


During Call Wait Time-     3.076433



**3. What is the number of calls taken for each time block?**



**North Branch Calls per time block:**

12:00 PM-    45

10:00 AM -   43

2:00 PM -    35

3:00 PM -    30

4:00 PM -    28

11:00 AM -   25

5:00 PM -    17

1:00 PM -    17

9:00 AM -     5

**South Branch Calls per time block:**

12:00 PM -   75

10:00 AM -   56

3:00 PM  -   43

11:00 AM -   31

2:00 PM  -   30

5:00 PM  -   26

4:00 PM -    25

1:00 PM -    23

9:00 AM  -    5



## Cleaning Data

Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets:

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them?
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [28]:
# Clean your data here

**Check for missing values**

In [29]:
north_missing = north_df.isna().sum()
north_missing

Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       163
During Call Wait Time      0
Calls                      0
dtype: int64

In [30]:
#northpct_missing = north_df.isna().mean()

#northpct_missing*100


In [31]:
south_missing = south_df.isna().sum()
south_missing

Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       188
During Call Wait Time      0
Calls                      0
dtype: int64

In [32]:
#southpct_missing=south_df.isna().mean()
#southpct_missing*100

In [33]:
#south_df.groupby('Incoming or Outgoing')['Incoming Wait Time'].value_counts(dropna=False)

# Counting occurences as well as missing values:

**count of missing values of a column by group:**

In [34]:
north_df.groupby(['Incoming or Outgoing'])['Incoming Wait Time'].apply(lambda x: x.isnull().sum())


Incoming or Outgoing
Incoming      0
Outgoing    163
Name: Incoming Wait Time, dtype: int64

In [35]:
south_df.groupby(['Incoming or Outgoing'])['Incoming Wait Time'].apply(lambda x: x.isnull().sum())


Incoming or Outgoing
Incoming      0
Outgoing    188
Name: Incoming Wait Time, dtype: int64

**All of the null values under incoming wait time correspond with outgoing calls. These values will be filled with 0**

In [36]:
north_df['Incoming Wait Time']=north_df['Incoming Wait Time'].fillna(0)

In [37]:
south_df['Incoming Wait Time']=south_df['Incoming Wait Time'].fillna(0)

In [38]:
#south_df.isna().sum()



**Verifying the filled values with groupby value counts**

In [39]:
south_df.groupby('Incoming or Outgoing')['Incoming Wait Time'].value_counts().unstack()

Incoming Wait Time      0.0   1.0   2.0   3.0   4.0   5.0
Incoming or Outgoing                                     
Incoming                NaN  20.0  28.0  29.0  30.0  19.0
Outgoing              188.0   NaN   NaN   NaN   NaN   NaN

In [40]:
north_df.groupby(['Incoming or Outgoing', 'Incoming Wait Time']).size().unstack(fill_value=0)

Incoming Wait Time    0.0  1.0  2.0  3.0  4.0  5.0
Incoming or Outgoing                              
Incoming                0   17   16   12   20   17
Outgoing              163    0    0    0    0    0

**Checking for Duplicate rows:**

In [41]:
northduplicate_rows = north_df[north_df.duplicated()]
print("North duplicated rows: ", northduplicate_rows.shape)

North duplicated rows:  (2, 14)


In [42]:
north_df[north_df.duplicated()]

Branch   Call Purpose Time Block Incoming or Outgoing Queue  Rep ID Sale  \
50   North  Sales Support   12:00 PM             Incoming     A     Joe   NO   
105  North  Sales Support   10:00 AM             Outgoing     A  Amanda   NO   

     Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
50           0          0            10                  1   
105          0          0            11                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
50                  4.0                      3      1  
105                 0.0                      2      1

In [43]:
southduplicate_rows = south_df[south_df.duplicated()]
print("South duplicated rows: ", southduplicate_rows.shape)

South duplicated rows:  (3, 14)


In [44]:
south_df[south_df.duplicated()]

Branch   Call Purpose Time Block Incoming or Outgoing Queue Rep ID Sale  \
58   South  Sales Support   12:00 PM             Incoming     D   Kate   NO   
170  South  Sales Support   11:00 AM             Outgoing     D  Randy   NO   
180  South  Sales Support   12:00 PM             Outgoing     C  Sandy   NO   

     Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
58           0          0             8                  1   
170          0          0             8                  3   
180          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
58                  3.0                      3      1  
170                 0.0                      2      1  
180                 0.0                      4      1

In [45]:
#pd.set_option('display.max_rows', None)

#north_df.duplicated()

#See boolean output for all duplicated rows

In [46]:
#Count duplicate in a DataFrame

north_df.duplicated().sum()

south_df.duplicated().sum()

2

3

## Make a copy of data sets and remove duplicated rows 

In [47]:
north_df_copy=north_df.copy()
south_df_copy=south_df.copy()

In [48]:
north_df_copy=north_df_copy.drop_duplicates()

In [49]:
south_df_copy=south_df_copy.drop_duplicates()

In [50]:
north_df_copy.duplicated().sum()

0

In [51]:
south_df_copy.duplicated().sum()

0

In [61]:
#north_df_copy.info()
#south_df_copy.info()

# Side-note-Get count of Missing values of rows  in pandas python:


Method 1: north_df.isnull().sum(axis = 1)**


Method 2: for i in range(len(north_df.index)) :
    print("Nan in row ", i , " : " ,  north_df.iloc[i].isnull().sum())

**Rectifying duplicate Yes values in North data set**

In [53]:
north_df_copy['Sale'].value_counts()

NO      232
YES      10
YES       1
Name: Sale, dtype: int64

In [54]:
north_df_copy['Sale']=north_df_copy['Sale'].str.strip()

In [55]:
north_df_copy['Sale'].value_counts()

NO     232
YES     11
Name: Sale, dtype: int64

**Used .strip to remove spaces at the beginning and at the end of the string. The values were the same in terms of casing, so the difference appeared to be within the leading and trailimg whitespaces, which ultimately proved to be correct.**

**Unnecessary data**

In [56]:
#North- columns with a high percentage of the same value
num_rows = len(north_df_copy.index)
low_information_cols = [] #

for col in north_df_copy.columns:
    cnts = north_df_copy[col].value_counts(dropna=False)
    top_pct = (cnts/num_rows).iloc[0]
    
    if top_pct > 0.95:
        low_information_cols.append(col)
        print('{0}: {1:.5f}%'.format(col, top_pct*100))
        print(cnts)
        print()

Branch: 100.00000%
North    243
Name: Branch, dtype: int64

Sale: 95.47325%
NO     232
YES     11
Name: Sale, dtype: int64

Lost Call: 97.94239%
0    238
1      5
Name: Lost Call, dtype: int64

Abandoned: 97.11934%
0    236
1      7
Name: Abandoned, dtype: int64

Calls: 100.00000%
1    243
Name: Calls, dtype: int64



In [57]:
#South- columns with a high percentage of the same value
num_rows = len(south_df_copy.index)
low_information_cols = [] #

for col in south_df_copy.columns:
    cnts = south_df_copy[col].value_counts(dropna=False)
    top_pct = (cnts/num_rows).iloc[0]
    
    if top_pct > 0.95:
        low_information_cols.append(col)
        print('{0}: {1:.5f}%'.format(col, top_pct*100))
        print(cnts)
        print()

Branch: 100.00000%
South    311
Name: Branch, dtype: int64

Sale: 98.07074%
NO     305
YES      6
Name: Sale, dtype: int64

Lost Call: 96.46302%
0    300
1     11
Name: Lost Call, dtype: int64

Abandoned: 98.71383%
0    307
1      4
Name: Abandoned, dtype: int64

Calls: 100.00000%
1    311
Name: Calls, dtype: int64



In [58]:
south_df_copy.shape
north_df_copy.shape

(311, 14)

(243, 14)

### Cleaning Data Results

Make note of your results!

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?

**All of the null values under the 'Incoming Wait Time' column corresponded with outgoing calls.This was verified by doing a value count of the two columns ('Incoming or Outgoing' and 'Incoming Wait Time') using the groupby operation. These values were filled with 0**

2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?

**The 2 different 'YES' values were caused by leading and trailing whitespaces. The .strip function was used to rectify this and make the values consistent.**

3. Are there duplicates in the two data sets? If there are, how did you handle them?

**After assigning the 'Unnamed' column as the index column, 2 duplicate rows were found in the North data set and 3 duplicate rows were found in the South data set. These were dropped to proceed with further analysis.**

4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

**No unnecessary data as of yet**

## Data Manipulation

Before you begin answering any questions, combine the two datasets together to create a third dataframe called `df`. You can use this third dataframe to compare the two call centers to company-wide trends.

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

In [59]:
# Manipulate data here

### Data Manipulation Results

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each company and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why.

1. What is the average abandonment rate per queue?
2. What is the service level and average speed of answer per each rep?
3. For each type of call purpose, how many calls are outgoing vs. incoming?

In [60]:
# Create visualization 1 here
# The dictionary abandonment_rates has the data you need.

abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

NameError: name 'df' is not defined

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

In [ ]:
# Create visualization 3 here
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

### Visualization Results

For each chart you created, explain why you chose the chart style you chose.

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.

1. Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?
2. Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?
3. Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

## Submit Your Work

After you have completed your work on the assignment, push your work to your Github repo. Navigate to the repo, copy the link to your assignment and paste the URL in the Canvas assignment's submission box. 

## Bonus Mission

Create a visualization that answers this question:

1. For each call purpose, how many calls (incoming and outgoing) take place in each time block?

In [ ]:
# Create your visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot:
